src/lib.rs

```diff
@@ -1,14 +1,180 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+mod args;
+mod context;
+mod rule;
+mod violation;
+
+use std::{
+    borrow::Cow,
+    process,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+pub use args::Args;
+use clap::Parser;
+use context::QueryMatchContext;
+use rule::{ResolvedRule, Rule, RuleBuilder, RuleListenerBuilder};
+use tree_sitter::Query;
+use violation::ViolationBuilder;
+
+use crate::context::Context;
+
+#[macro_export]
+macro_rules! regex {
+    ($re:expr $(,)?) => {{
+        static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
+        RE.get_or_init(|| regex::Regex::new($re).unwrap())
+    }};
+}
+
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP: &str = "_tree_sitter_lint_capture";
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT: &str = "@_tree_sitter_lint_capture";
+
+pub fn run(_args: Args) {
+    let language = tree_sitter_rust::language();
+    let context = Context::new(language);
+    let resolved_rules = get_rules()
```

Comment:
Rules/listeners get "resolved" by eg calling the `create()` callback and then actually parsing the tree-sitter queries etc